src/mame/olivetti/att6300p.cpp

```diff
@@ -0,0 +1,1050 @@
+    control on a 1K page basis.  Additionally, IO accesses can be virtualized,
+    on a per-port basis, allowing device ownership and access control, and
+    even the implemenation of virtual devices.  In effect, this system was the
+    first to support a hardware virtual 8086 mode.
+
+    TODO: while an attempt has been made to emulate the virtualization
+    hardware, it is mostly unverified, and will almost certainly not work.
+    In order to verify it, someone needs to unearth a copy of the AT&T Unix
+    release that was developed and sold specifically for the 6300+, in
+    addition to the OS Merge (AKA Simul-Task) software that's needed to run
+    Unix and DOS concurrently.  There are seven disks labelled "UNIX System V
+    Release 2.0 Foundation Set" and one labelled "UNIX System V Release 2.0
+    Simul-Task OS Merge Disk".  All disks are also labelled with "AT&T
+    Personal Computer 6300 Plus".  Anyone with a disk set to donate or sell
+    is kindly requested to contact donohoe00 on the VCFed forum or github.
+
+    Some of the enhanced functionality, such as switching in and out of
+    protected mode, has been verified (since the BIOS does this in order to
+    test memory above 640K).  This involves external logic to reset the CPU
+    (the only way to return to Real Mode) and to enable/disable address
+    lines A20-A23.
+
+    Sources:
+
+    https://archive.org/details/att_hardware_reference_manual_att_personal_computer_6300_plus/page/n41/mode/2up
+    https://bitsavers.org/pdf/att/6300/ATT_Personal_Computer_6300_Plus_-_ROM-BIOS_Listing_-_Rel_2.05_Issue_1_-_1986.pdf
+
+****************************************************************************/
+
+#include "emu.h"
+
+#include "att6300p_mmu.h"
+
+#include "bus/isa/isa.h"
+#include "bus/isa/isa_cards.h"
+#include "cpu/i86/i286.h"
+#include "cpu/mcs48/mcs48.h"
+#include "imagedev/floppy.h"
+#include "machine/am9517a.h"
+#include "machine/z80ctc.h"
+#include "m24_kbd.h"
+#include "machine/mm58274c.h"
+#include "machine/pit8253.h"
+#include "machine/pic8259.h"
+#include "machine/ram.h"
+#include "sound/spkrdev.h"
+#include "speaker.h"
+
+#include "formats/naslite_dsk.h"
+#include "formats/m20_dsk.h"
+
+#include "softlist_dev.h"
+
+
+namespace {
+
+class att6300p_state : public driver_device
+{
+public:
+	att6300p_state(const machine_config &mconfig, device_type type, const char *tag) :
+		driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_ram(*this, RAM_TAG),
+		m_isabus(*this, "isabus"),
+		m_dmac(*this, "dmac"),
+		m_pic(*this, "pic"),
+		m_pit(*this, "pit"),
+		m_speaker(*this, "speaker"),
+		m_kbc(*this, "kbc"),
+		m_keyboard(*this, "keyboard"),
+		m_ctc(*this, "ctc"),
+		m_mmu(*this, "mmu"),
+		m_dsw(*this, "DSW"),
+		m_map_rom(*this, "addrmap"),
+		m_nmi_enable(false)
+	{ }
+
+	void att6300p(machine_config &config);
+
+protected:
+	virtual void machine_start() override ATTR_COLD;
+	virtual void machine_reset() override ATTR_COLD;
+
+private:
+	/*
```

Comment:
This is really quite verbose. Maybe it can moved down to where it's used and be a bit more condensed (assuming this isn't reverse engineered but also mentioned in the manual).